Add explicit type annotations to debug asserts in Bevy ECS

## Bevy version

0.15.1 (I've tried 0.15.2, it happens with that version too.)

## \[Optional\] Relevant system information

cargo 1.87.0-nightly (ce948f461 2025-02-14)
rustc 1.87.0-nightly

I also tried:

cargo 1.84.1 (66221abde 2024-11-19)
rustc 1.84.1 (e71f9a9a9 2025-01-27)

## What you did

I tried building a crate that has both `bevy_ecs` and `serde_json` as a dependency for `wasm32-unknown-unknown`.

## What went wrong

The Rust compiler seems to get confused by the fact that both `usize` and `serde_json::value::Value` implement `From<usize>`.
For some reason I've only experienced that when building for `wasm32-unknown-unknown`. Building for Linux, Windows or Mac OS doesn't seem to run into this. I really don't know why.

```
error[E0283]: type annotations needed
   --> /home/robin/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/bevy_ecs-0.15.2/src/storage/blob_array.rs:251:58
    |
251 |         debug_assert_eq!(self.capacity, current_capacity.into());
    |                                                          ^^^^
    |
    = note: cannot satisfy `_: From<NonZero<usize>>`
    = note: required for `NonZero<usize>` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
    |
251 -         debug_assert_eq!(self.capacity, current_capacity.into());
251 +         debug_assert_eq!(self.capacity, <NonZero<usize> as Into<T>>::into(current_capacity));
    |

error[E0283]: type annotations needed
   --> /home/robin/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/bevy_ecs-0.15.2/src/storage/blob_array.rs:251:58
    |
251 |         debug_assert_eq!(self.capacity, current_capacity.into());
    |         -------------------------------------------------^^^^--- type must be known at this point
    |
    = note: multiple `impl`s satisfying `usize: PartialEq<_>` found in the following crates: `core`, `serde_json`:
            - impl PartialEq for usize;
            - impl PartialEq<serde_json::value::Value> for usize;
help: try using a fully qualified path to specify the expected types
    |
251 -         debug_assert_eq!(self.capacity, current_capacity.into());
251 +         debug_assert_eq!(self.capacity, <NonZero<usize> as Into<T>>::into(current_capacity));

error[E0283]: type annotations needed
  --> /home/robin/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/bevy_ecs-0.15.2/src/storage/thin_array_ptr.rs:85:52
   |
85 |         assert_eq!(self.capacity, current_capacity.into());
   |                                                    ^^^^
   |
   = note: cannot satisfy `_: From<NonZero<usize>>`
   = note: required for `NonZero<usize>` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
   |
85 -         assert_eq!(self.capacity, current_capacity.into());
85 +         assert_eq!(self.capacity, <NonZero<usize> as Into<T>>::into(current_capacity));
   |

error[E0283]: type annotations needed
  --> /home/robin/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/bevy_ecs-0.15.2/src/storage/thin_array_ptr.rs:85:52
   |
85 |         assert_eq!(self.capacity, current_capacity.into());
   |         -------------------------------------------^^^^--- type must be known at this point
   |
   = note: multiple `impl`s satisfying `usize: PartialEq<_>` found in the following crates: `core`, `serde_json`:
           - impl PartialEq for usize;
           - impl PartialEq<serde_json::value::Value> for usize;
help: try using a fully qualified path to specify the expected types
   |
85 -         assert_eq!(self.capacity, current_capacity.into());
85 +         assert_eq!(self.capacity, <NonZero<usize> as Into<T>>::into(current_capacity));
   |
```

--- a/crates/bevy_ecs/src/storage/blob_array.rs
+++ b/crates/bevy_ecs/src/storage/blob_array.rs
@@ -256,7 +256,7 @@ impl BlobArray {
         new_capacity: NonZeroUsize,
     ) {
         #[cfg(debug_assertions)]
-        debug_assert_eq!(self.capacity, current_capacity.into());
+        debug_assert_eq!(self.capacity, usize::from(current_capacity));
         if !self.is_zst() {
             // SAFETY: `new_capacity` can't overflow usize
             let new_layout =

--- a/crates/bevy_ecs/src/storage/blob_array.rs
+++ b/crates/bevy_ecs/src/storage/blob_array.rs
@@ -256,7 +256,7 @@ impl BlobArray {
         new_capacity: NonZeroUsize,
     ) {
         #[cfg(debug_assertions)]
-        debug_assert_eq!(self.capacity, usize::from(current_capacity));
+        debug_assert_eq!(self.capacity, current_capacity.get());
         if !self.is_zst() {
             // SAFETY: `new_capacity` can't overflow usize
             let new_layout =

--- a/crates/bevy_ecs/src/storage/thin_array_ptr.rs
+++ b/crates/bevy_ecs/src/storage/thin_array_ptr.rs
@@ -87,7 +87,7 @@ impl<T> ThinArrayPtr<T> {
     /// - The caller should update their saved `capacity` value to reflect the fact that it was changed
     pub unsafe fn realloc(&mut self, current_capacity: NonZeroUsize, new_capacity: NonZeroUsize) {
         #[cfg(debug_assertions)]
-        assert_eq!(self.capacity, usize::from(current_capacity));
+        assert_eq!(self.capacity, current_capacity.get());
         self.set_capacity(new_capacity.get());
         if size_of::<T>() != 0 {
             let new_layout =

--- a/crates/bevy_ecs/src/storage/thin_array_ptr.rs
+++ b/crates/bevy_ecs/src/storage/thin_array_ptr.rs
@@ -87,7 +87,7 @@ impl<T> ThinArrayPtr<T> {
     /// - The caller should update their saved `capacity` value to reflect the fact that it was changed
     pub unsafe fn realloc(&mut self, current_capacity: NonZeroUsize, new_capacity: NonZeroUsize) {
         #[cfg(debug_assertions)]
-        assert_eq!(self.capacity, current_capacity.into());
+        assert_eq!(self.capacity, usize::from(current_capacity));
         self.set_capacity(new_capacity.get());
         if size_of::<T>() != 0 {
             let new_layout =

